File: sdks/rust/src/capture.rs
//! The capture surface: what a caller hands the SDK per LLM call, and the
//! (worker-side) conversion into wire records.
//!
//! Building an [`LlmCall`] and calling `Client::record` is the only thing that
//! happens on the live request path — it must stay a cheap move into a buffer.
//! All of the work here (redaction, hashing, id derivation) runs later, on the
//! background worker, off the hot path.

use crate::config::{Config, RedactionPolicy};
use crate::redact;
use crate::wire::{
    self, cap_metadata, EventKind, GitContext, IngestBatch, RawEvent, TokenUsage, ToolCallStatus,
    ToolCallWire,
};
use chrono::{DateTime, Utc};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::time::Duration;

/// The excerpt cap for the standard (non-raw) path, in Unicode scalars.
const EXCERPT_MAX_CHARS: usize = 320;

/// One captured tool invocation. The SDK is in the call path, so it has the
/// real args and result — it hashes/sizes them here (never ships them raw).
#[derive(Debug, Clone)]
pub struct ToolCallInput {
    /// Bare tool name (`Bash`, `create_pr`).
    pub name: String,
    /// `builtin` or `mcp:<server>`.
    pub server: String,
    /// The call's arguments, if any. Hashed and sized; never shipped.
    pub args: Option<serde_json::Value>,
    /// Byte length of the result/output (the SDK sizes it; never ships it).
    pub result_bytes: u32,
    pub status: ToolCallStatus,
    pub started_at: DateTime<Utc>,
    pub ended_at: Option<DateTime<Utc>>,
    /// Allowlisted command verbs for shell-ish tools (≤3, each ≤40 chars).
    pub command_families: Vec<String>,
}

impl ToolCallInput {
    /// A minimal builtin tool call.
    #[must_use]
    pub fn new(name: impl Into<String>, status: ToolCallStatus) -> Self {
        Self {
            name: name.into(),
            server: "builtin".into(),
            args: None,
            result_bytes: 0,
            status,
            started_at: Utc::now(),
            ended_at: None,
            command_families: Vec::new(),
        }
    }
}

/// One captured LLM call. Construct with [`LlmCall::new`] and set the fields you
/// have; `prompt`/`completion` are raw here and are redacted on the worker.
#[derive(Debug, Clone)]
pub struct LlmCall {
    pub provider: String,
    pub model: Option<String>,
    /// Trace/conversation id used to group calls into a session downstream.
    pub session_id: String,
    pub kind: EventKind,
    pub tokens: TokenUsage,
    pub started_at: DateTime<Utc>,
    /// When the first piece of the model's output arrived, when the caller
    /// watched for it (a streamed response). Left `None` on a call that returns
    /// in one piece — there is no first chunk to have seen.
    pub first_token_at: Option<DateTime<Utc>>,
    pub duration: Option<Duration>,
    pub prompt: Option<String>,
    pub completion: Option<String>,
    pub cwd: Option<String>,
    pub git: Option<GitContext>,
    pub tool_calls: Vec<ToolCallInput>,
    /// Per-call attribution tags. Highest-priority layer: these override
    /// `Config::metadata` for any shared key. Capped before send.
    pub metadata: BTreeMap<String, String>,
}

impl LlmCall {
    /// A call with `started_at = now` and `kind = assistant_message`.
    #[must_use]
    pub fn new(provider: impl Into<String>, session_id: impl Into<String>) -> Self {
        Self {
            provider: provider.into(),
            model: None,
            session_id: session_id.into(),
            kind: EventKind::AssistantMessage,
            tokens: TokenUsage::default(),
            started_at: Utc::now(),
            first_token_at: None,
            duration: None,
            prompt: None,
            completion: None,
            cwd: None,
            git: None,
            // Unknown until the caller says otherwise: an SDK user who never
            // touches this field has not told us how they pay, and guessing is
            // what put $39.65 of imaginary spend on a subscription account.
            tool_calls: Vec::new(),
            metadata: BTreeMap::new(),
        }
    }

    /// Set the model.
    #[must_use]
    pub fn model(mut self, model: impl Into<String>) -> Self {
        self.model = Some(model.into());
        self
    }

    /// Set token usage.
    #[must_use]
    pub fn tokens(mut self, tokens: TokenUsage) -> Self {
        self.tokens = tokens;
        self
    }

    /// State when the first piece of the model's output arrived. Call this from
    /// a streaming loop, on the first chunk; leave it unset otherwise.
    #[must_use]
    pub fn first_token_at(mut self, at: DateTime<Utc>) -> Self {
        self.first_token_at = Some(at);
        self
    }

    /// Set the prompt and completion text (raw; redacted on the worker).
    #[must_use]
    pub fn text(mut self, prompt: impl Into<String>, completion: impl Into<String>) -> Self {
        self.prompt = Some(prompt.into());
        self.completion = Some(completion.into());
        self
    }

    /// Attach a single per-call attribution tag (overwriting any previous value
    /// for `key`). Per-call tags override [`Config::metadata`] for shared keys.
    #[must_use]
    pub fn metadata(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.metadata.insert(key.into(), value.into());
        self
    }

    /// Merge a map of per-call attribution tags (each overwriting any previous
    /// value for the same key).
    #[must_use]
    pub fn with_metadata<K, V>(mut self, tags: impl IntoIterator<Item = (K, V)>) -> Self
    where
        K: Into<String>,
        V: Into<String>,
    {
        for (k, v) in tags {
            self.metadata.insert(k.into(), v.into());
        }
        self
    }
}

/// Truncate to at most `max` Unicode scalars, appending an elision marker.
fn truncate_chars(s: &str, max: usize) -> String {
    if s.chars().count() <= max {
        return s.to_string();
    }
    let head: String = s.chars().take(max).collect();
    format!("{head}…")
}

/// sha256 hex of `bytes`.
fn sha256_hex(bytes: &[u8]) -> String {
    let mut h = Sha256::new();
    h.update(bytes);
    format!("{:x}", h.finalize())
}

/// Build the privacy-reduced `(args_hash, signature_hash, args_bytes)` triple
/// for a tool call's arguments.
fn hash_args(args: Option<&serde_json::Value>) -> (String, String, u32) {
    let Some(args) = args else {
        return (String::new(), "none".into(), 0);
    };
    let serialized = serde_json::to_string(args).unwrap_or_default();
    let args_hash = sha256_hex(serialized.as_bytes());
    let signature = match args {
        serde_json::Value::Object(map) => {
            let mut keys: Vec<&str> = map.keys().map(String::as_str).collect();
            keys.sort_unstable();
            sha256_hex(keys.join(",").as_bytes())
        }
        _ => "none".into(),
    };
    (args_hash, signature, serialized.len() as u32)
}

/// Convert one captured call into a wire event plus its tool-call records.
fn event_from_call(cfg: &Config, call: &LlmCall, seq: u64) -> (RawEvent, Vec<ToolCallWire>) {
    let source_ref = format!(
        "{}::{}::{}",
        call.session_id,
        call.started_at.timestamp_millis(),
        seq
    );
    let source_event_id = wire::source_event_id(&cfg.device_id, &source_ref);

    let content_excerpt = build_excerpt(cfg, call);
    let metadata = resolve_metadata(cfg, call);

    let event = RawEvent {
        source_event_id: source_event_id.clone(),
        ts: call.started_at,
        // The SDK held the clock for this call, so it states the span's ends
        // rather than leaving a reader to reconstruct them from `ts` and a
        // duration that may not be set.
        started_at: Some(call.started_at),
        first_token_at: call.first_token_at,
        kind: call.kind,
        agent: cfg.agent.clone(),
        provider: call.provider.clone(),
        model: call.model.clone(),
        session_id: call.session_id.clone(),
        tokens: call.tokens,
        cwd: call.cwd.clone(),
        git: call.git.clone(),
        duration_ms: call.duration.map(|d| d.as_millis() as u64),
        content_excerpt,
        metadata,
    };

    let tool_calls = call
        .tool_calls
        .iter()
        .enumerate()
        .map(|(i, tc)| {
            let (args_hash, signature_hash, args_bytes) = hash_args(tc.args.as_ref());
            let external_call_id = format!(
                "tc_{}",
                &wire::content_hash(&[&source_event_id, &i.to_string()])[..16]
            );
            ToolCallWire {
                external_call_id,
                session_id: call.session_id.clone(),
                source_event_id: source_event_id.clone(),
                segment_id: None,
                agent: cfg.agent.clone(),
                server: tc.server.clone(),
                name: tc.name.clone(),
                turn_index: None,
                call_index: i as u32,
                started_at: tc.started_at,
                ended_at: tc.ended_at,
                status: tc.status,
                args_hash,
                signature_hash,
                args_bytes,
                result_bytes: tc.result_bytes,
                model: call.model.clone(),
                command_families: tc.command_families.iter().take(3).cloned().collect(),
            }
        })
        .collect();

    (event, tool_calls)
}

/// Resolve the per-event metadata: `Config` defaults are the base layer and
/// per-call tags override them (per-call wins on a shared key), then the caps
/// are applied. Returns `None` when the merged map is empty so the wire key is
/// omitted. (The Rust SDK has no ambient layer — see the crate README.)
fn resolve_metadata(cfg: &Config, call: &LlmCall) -> Option<BTreeMap<String, String>> {
    if cfg.metadata.is_empty() && call.metadata.is_empty() {
        return None;
    }
    let mut merged = cfg.metadata.clone();
    // Per-call wins: inserting overwrites any default with the same key.
    for (k, v) in &call.metadata {
        merged.insert(k.clone(), v.clone());
    }
    let capped = cap_metadata(merged);
    if capped.is_empty() {
        None
    } else {
        Some(capped)
    }
}

/// Build the redacted excerpt from a call's prompt + completion, honoring the
/// configured redaction policy and (for the standard path) the 320-char cap.
fn build_excerpt(cfg: &Config, call: &LlmCall) -> Option<String> {
    let mut joined = String::new();
    if let Some(p) = &call.prompt {
        joined.push_str(p);
    }
    if let Some(c) = &call.completion {
        if !joined.is_empty() {
            joined.push_str("\n---\n");
        }
        joined.push_str(c);
    }
    if joined.is_empty() {
        return None;
    }

    let scrubbed = match cfg.redaction {
        RedactionPolicy::Floor => redact::redact(&joined).text,
        RedactionPolicy::None => joined,
    };

    // Raw mode ships the full (redacted) turns for server-side summarization;
    // the standard path caps the excerpt.
    if cfg.sends_full_turns() {
        Some(scrubbed)
    } else {
        Some(truncate_chars(&scrubbed, EXCERPT_MAX_CHARS))
    }
}

/// Drain a batch of captured calls into a wire [`IngestBatch`]. `seq` is a
/// monotonic counter used to keep per-call dedupe keys distinct within a run.
pub(crate) fn build_batch(
    cfg: &Config,
    calls: impl IntoIterator<Item = LlmCall>,
    seq: &mut u64,
) -> IngestBatch {
    let mut events = Vec::new();
    let mut tool_calls = Vec::new();
    let mut source_ids = Vec::new();
    // Each session named to its account — the app — under the provider its
    // events actually used, so attribution happens AT INGEST, per app, like a
    // daemon-observed login. A session's calls share one provider in practice;
    // if they ever didn't, last-write-wins matches the map semantics.
    let mut session_installs = std::collections::BTreeMap::new();

    for call in calls {
        *seq += 1;
        if !call.provider.is_empty() {
            session_installs.insert(
                call.session_id.clone(),
                wire::SessionInstall {
                    provider_account_id: cfg.app.clone(),
                    provider: call.provider.clone(),
                },
            );
        }
        let (event, tcs) = event_from_call(cfg, &call, *seq);
        source_ids.push(event.source_event_id.clone());
        tool_calls.extend(tcs);
        events.push(event);
    }

    IngestBatch {
        batch_id: wire::batch_id(&source_ids),
        device_id: cfg.device_id.clone(),
        daemon_version: cfg.client_version.clone(),
        app: Some(cfg.app.clone()),
        events,
        tool_calls,
        session_installs,
        // Always send an explicit value reflecting the config so backend usage
        // is off-by-default but users can opt in.
        auto_taxonomy: Some(cfg.auto_taxonomy),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::Config;

    fn cfg() -> Config {
        Config::new("msk_test", "raw_sdk_openai", "test-app").with_device_id("dev_test")
    }

    #[test]
    fn build_batch_redacts_excerpt_and_caps_length() {
        let mut seq = 0;
        let call = LlmCall::new("openai", "sess_1").model("gpt-x").text(
            "here is my key sk-ant-0123456789abcdefghijABCDEF",
            "ok done",
        );
        let batch = build_batch(&cfg(), [call], &mut seq);

        assert_eq!(batch.events.len(), 1);
        let ev = &batch.events[0];
        assert_eq!(ev.agent, "raw_sdk_openai");
        assert_eq!(ev.provider, "openai");
        let excerpt = ev.content_excerpt.as_ref().unwrap();
        assert!(
            excerpt.contains("[REDACTED:anthropic_key]"),
            "got {excerpt:?}"
        );
        assert!(!excerpt.contains("sk-ant-0123"));
        assert!(excerpt.chars().count() <= EXCERPT_MAX_CHARS + 1);
        assert!(ev.source_event_id.starts_with("evt_"));
        assert!(batch.batch_id.starts_with("batch_"));
    }

    /// The SDK is in the call path, so it can state the span's ends. `started_at`
    /// always ships (it is `ts`'s own provenance made explicit); the first-token
    /// instant ships only when the caller watched a stream and said so, because a
    /// call that returns in one piece never had a first chunk to time.
    #[test]
    fn the_call_states_the_instants_it_saw_and_omits_the_one_it_did_not() {
        let mut seq = 0;
        let quiet = LlmCall::new("openai", "sess_1");
        let started = quiet.started_at;
        let streamed =
            LlmCall::new("openai", "sess_2").first_token_at(started + Duration::from_millis(140));

        let batch = build_batch(&cfg(), [quiet, streamed], &mut seq);
        let (a, b) = (&batch.events[0], &batch.events[1]);

        assert_eq!(a.started_at, Some(started));
        assert_eq!(
            a.ts, started,
            "ts is unchanged — the new field sits beside it"
        );
        assert_eq!(a.first_token_at, None, "no stream, no first chunk to time");
        assert_eq!(
            b.first_token_at,
            Some(started + Duration::from_millis(140)),
            "the instant the caller stated survives verbatim"
        );

        // Additive on the wire: absent means absent, never null.
        let json = serde_json::to_value(a).unwrap();
        assert!(json.get("first_token_at").is_none());
        assert!(json.get("started_at").is_some());
    }

    #[test]
    fn tool_calls_carry_hashes_not_raw_args() {
        let mut seq = 0;
        let mut call = LlmCall::new("anthropic", "sess_1");
        call.tool_calls.push(ToolCallInput {
            name: "Bash".into(),
            server: "builtin".into(),
            args: Some(serde_json::json!({ "command": "rm -rf /tmp/secret", "timeout": 5 })),
            result_bytes: 128,
            status: ToolCallStatus::Success,
            started_at: Utc::now(),
            ended_at: None,
            command_families: vec!["rm".into()],
        });
        let batch = build_batch(&cfg(), [call], &mut seq);

        assert_eq!(batch.tool_calls.len(), 1);
        let tc = &batch.tool_calls[0];
        assert_eq!(tc.name, "Bash");
        assert_eq!(
            tc.args_bytes,
            serde_json::to_string(
                &serde_json::json!({ "command": "rm -rf /tmp/secret", "timeout": 5 })
            )
            .unwrap()
            .len() as u32
        );
        assert_eq!(tc.args_hash.len(), 64); // sha256 hex
        assert_ne!(tc.signature_hash, "none");
        // The raw command must never appear anywhere in the serialized batch.
        let json = serde_json::to_string(&batch).unwrap();
        assert!(
            !json.contains("rm -rf /tmp/secret"),
            "raw args leaked into wire"
        );
    }

    #[test]
    fn auto_taxonomy_defaults_off_and_opts_in() {
        // Default config: taxonomy off → explicit `auto_taxonomy: false`.
        let mut seq = 0;
        let batch = build_batch(&cfg(), [LlmCall::new("openai", "sess_1")], &mut seq);
        assert_eq!(batch.auto_taxonomy, Some(false));
        let j: serde_json::Value = serde_json::to_value(&batch).unwrap();
        assert_eq!(j["auto_taxonomy"], false);

        // Opt in: flag true → wire `auto_taxonomy: true`.
        let mut on = cfg();
        on.auto_taxonomy = true;
        let mut seq = 0;
        let batch = build_batch(&on, [LlmCall::new("openai", "sess_1")], &mut seq);
        assert_eq!(batch.auto_taxonomy, Some(true));
        let j: serde_json::Value = serde_json::to_value(&batch).unwrap();
        assert_eq!(j["auto_taxonomy"], true);
    }

    #[test]
    fn metadata_merges_with_per_call_winning_and_omits_when_empty() {
        // No metadata anywhere → key omitted on the wire.
        let mut seq = 0;
        let batch = build_batch(&cfg(), [LlmCall::new("openai", "sess_1")], &mut seq);
        assert!(batch.events[0].metadata.is_none());
        let j: serde_json::Value = serde_json::to_value(&batch).unwrap();
        assert!(j["events"][0].get("metadata").is_none());

        // Config defaults < per-call: per-call `feature` overrides the default,
        // the default-only `environment` survives, the per-call-only `team` is
        // added.
        let cfg = cfg()
            .with_metadata("environment", "prod")
            .with_metadata("feature", "default_feature");
        let call = LlmCall::new("openai", "sess_1")
            .metadata("feature", "search")
            .with_metadata([("team", "growth")]);
        let mut seq = 0;
        let batch = build_batch(&cfg, [call], &mut seq);
        let md = batch.events[0].metadata.as_ref().unwrap();
        assert_eq!(md.get("environment").map(String::as_str), Some("prod"));
        assert_eq!(md.get("feature").map(String::as_str), Some("search"));
        assert_eq!(md.get("team").map(String::as_str), Some("growth"));
        // It serializes as a flat object.
        let j: serde_json::Value = serde_json::to_value(&batch).unwrap();
        assert_eq!(j["events"][0]["metadata"]["feature"], "search");
    }

    #[test]
    fn metadata_caps_excess_keys_by_sorted_order() {
        // 20 default keys "k00".."k19"; only the 16 smallest survive (k00..k15).
        let mut over = cfg();
        for i in 0..20 {
            over = over.with_metadata(format!("k{i:02}"), "v");
        }
        let mut seq = 0;
        let batch = build_batch(&over, [LlmCall::new("openai", "sess_1")], &mut seq);
        let md = batch.events[0].metadata.as_ref().unwrap();
        assert_eq!(md.len(), 16);
        assert!(md.contains_key("k15"));
        assert!(!md.contains_key("k16"));
    }

    #[test]
    fn metadata_truncates_over_long_values_before_send() {
        let call = LlmCall::new("openai", "sess_1").metadata("big", "v".repeat(500));
        let mut seq = 0;
        let batch = build_batch(&cfg(), [call], &mut seq);
        let md = batch.events[0].metadata.as_ref().unwrap();
        assert_eq!(md.get("big").unwrap().chars().count(), 256);
    }

    #[test]
    fn raw_mode_sends_full_untruncated_turns_still_floor_redacted() {
        let cfg = Config::new("msk", "raw_sdk_openai", "test-app")
            .with_device_id("dev_test")
            .with_remote("https://api.modelstat.ai", true);
        let mut seq = 0;
        let long = "word ".repeat(200); // > 320 chars
        let call = LlmCall::new("openai", "sess_1").text(long.clone(), "AKIAIOSFODNN7EXAMPLE");
        let batch = build_batch(&cfg, [call], &mut seq);
        let excerpt = batch.events[0].content_excerpt.as_ref().unwrap();
        assert!(
            excerpt.chars().count() > EXCERPT_MAX_CHARS,
            "raw mode must not truncate"
        );
        assert!(
            excerpt.contains("[REDACTED:aws_access_key]"),
            "floor still applies in raw mode"
        );
    }
}

File: sdks/node/src/capture.ts
/**
 * The capture surface: what a caller hands the SDK per LLM call, and the
 * (worker-side) conversion into wire records.
 *
 * Building an {@link LlmCall} and calling `Client.record` is the only thing
 * that happens on the live request path — it must stay a cheap push into a
 * buffer. All of the work here (redaction, hashing, id derivation) runs later,
 * on the background worker, off the hot path.
 */

import { sha256 } from "@noble/hashes/sha2.js";
import { bytesToHex, utf8ToBytes } from "@noble/hashes/utils.js";
import type { Config } from "./config.js";
import { ambientMetadata } from "./context.js";
import { redact } from "./redact.js";
import {
  batchId,
  capMetadata,
  contentHash,
  sourceEventId,
  zeroTokens,
  type Metadata,
  type EventKind,
  type GitContext,
  type IngestBatch,
  type RawEvent,
  type TokenUsage,
  type ToolCallStatus,
  type ToolCallWire,
} from "./wire.js";

/** The excerpt cap for the standard (non-raw) path, in Unicode code points. */
const EXCERPT_MAX_CHARS = 320;

/**
 * One captured tool invocation. The SDK is in the call path, so it has the
 * real args and result — it hashes/sizes them here (never ships them raw).
 */
export interface ToolCallInput {
  /** Bare tool name (`Bash`, `create_pr`). */
  name: string;
  /** `builtin` or `mcp:<server>`. Defaults to `builtin`. */
  server?: string;
  /** The call's arguments, if any. Hashed and sized; never shipped. */
  args?: unknown;
  /** Byte length of the result/output (the SDK sizes it; never ships it). */
  resultBytes?: number;
  status: ToolCallStatus;
  /** Defaults to now. */
  startedAt?: Date;
  endedAt?: Date;
  /** Allowlisted command verbs for shell-ish tools (≤3, each short). */
  commandFamilies?: string[];
}

/**
 * One captured LLM call. Construct with `new LlmCall(provider, sessionId)` and
 * either set the public fields directly or chain the fluent builders
 * `.model()`, `.tokens()`, `.text()`. `prompt`/`completion` are raw here and
 * are redacted on the worker.
 *
 * The model and token usage are exposed through the fluent builder *methods*
 * `.model(...)` / `.tokens(...)` (returning `this`); their backing values live
 * in the public `modelName` / `tokenUsage` fields, which you may also set or
 * read directly. (TypeScript can't share one name between a method and a data
 * field, so the Rust reference's `pub model` field + `.model()` builder is
 * split into a field with a distinct name plus a same-named builder.)
 */
export class LlmCall {
  provider: string;
  /** The model name, if known. Set via `.model(...)` or assigned directly. */
  modelName?: string;
  /** Trace/conversation id used to group calls into a session downstream. */
  sessionId: string;
  kind: EventKind = "assistant_message";
  /** Token usage. Set via `.tokens(...)` or assigned directly. */
  tokenUsage: TokenUsage = zeroTokens();
  startedAt: Date = new Date();
  durationMs?: number;
  prompt?: string;
  completion?: string;
  cwd?: string;
  git?: GitContext;
  toolCalls: ToolCallInput[] = [];
  /**
   * Per-call attribution tags. The highest-priority layer: these override the
   * ambient context layer and `Config` defaults on shared keys. Set via
   * `.metadata({...})` (which merges) or assigned directly. Capped before send.
   */
  metadataTags: Metadata = {};
  /**
   * Snapshot of the ambient ({@link withMetadata}) tags captured on the hot
   * path at `record()` time — *not* set by callers. The worker fills this in
   * because the merge runs later, off the `withMetadata` scope. The middle
   * layer: above `Config` defaults, below {@link LlmCall.metadataTags}.
   */
  ambientMetadataSnapshot?: Metadata;

  /** A call with `startedAt = now` and `kind = "assistant_message"`. */
  constructor(provider: string, sessionId: string) {
    this.provider = provider;
    this.sessionId = sessionId;
  }

  /** Fluent: set the model. Returns `this` for chaining. */
  model(model: string): this {
    this.modelName = model;
    return this;
  }

  /**
   * Fluent: set token usage (partial — any classes you omit default to zero).
   * Returns `this` for chaining.
   */
  tokens(tokens: Partial<TokenUsage>): this {
    this.tokenUsage = { ...zeroTokens(), ...tokens };
    return this;
  }

  /** Fluent: set the prompt and completion text (raw; redacted on the worker). */
  text(prompt: string, completion: string): this {
    this.prompt = prompt;
    this.completion = completion;
    return this;
  }

  /**
   * Fluent: merge per-call attribution tags (each key overwrites any previous
   * value, including the same-keyed default/ambient tag). Returns `this` for
   * chaining.
   *
   * The backing field is `metadataTags` (TypeScript can't share one name
   * between this builder method and a data field), which you may also set or
   * read directly.
   */
  metadata(tags: Metadata): this {
    this.metadataTags = { ...this.metadataTags, ...tags };
    return this;
  }
}

/** sha256 hex of `bytes`. */
function sha256Hex(bytes: Uint8Array): string {
  return bytesToHex(sha256(bytes));
}

/** Truncate to at most `max` Unicode code points, appending an elision marker. */
function truncateChars(s: string, max: number): string {
  // `Array.from` / spread iterates by code point, so surrogate pairs count as
  // one — matching Rust's `chars()` (Unicode scalar values).
  const points = Array.from(s);
  if (points.length <= max) {
    return s;
  }
  return points.slice(0, max).join("") + "…";
}

/**
 * Build the privacy-reduced `[argsHash, signatureHash, argsBytes]` triple for a
 * tool call's arguments.
 *
 * - `argsHash`  = sha256 hex of the canonical (compact) JSON of `args`, or `""`
 *   when there are no args.
 * - `signatureHash` = sha256 hex of the sorted top-level arg key names joined
 *   by `,`; the literal `"none"` when args are absent or not a plain object.
 * - `argsBytes` = UTF-8 byte length of that JSON.
 */
function hashArgs(args: unknown): [string, string, number] {
  if (args === undefined || args === null) {
    return ["", "none", 0];
  }
  // Compact JSON (no spaces) mirrors Rust's `serde_json::to_string`.
  const serialized = JSON.stringify(args) ?? "";
  const serializedBytes = utf8ToBytes(serialized);
  const argsHash = sha256Hex(serializedBytes);

  let signature: string;
  if (isPlainObject(args)) {
    const keys = Object.keys(args).sort();
    signature = sha256Hex(utf8ToBytes(keys.join(",")));
  } else {
    // Arrays, numbers, strings, booleans: no top-level key names.
    signature = "none";
  }
  return [argsHash, signature, serializedBytes.length];
}

/** Whether `v` is a plain JSON object (not an array, null, or primitive). */
function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/**
 * Format a `Date` as RFC3339 UTC with millisecond precision, e.g.
 * `"2026-06-19T00:00:00.000Z"`. `Date.toISOString()` already produces exactly
 * this shape.
 */
function rfc3339(d: Date): string {
  return d.toISOString();
}

/**
 * Build the redacted excerpt from a call's prompt + completion, honoring the
 * configured redaction policy and (for the standard path) the 320-char cap.
 * Returns `undefined` when there is no text to send.
 */
export function buildExcerpt(cfg: Config, call: LlmCall): string | undefined {
  let joined = "";
  if (call.prompt !== undefined) {
    joined += call.prompt;
  }
  if (call.completion !== undefined) {
    if (joined.length > 0) {
      joined += "\n---\n";
    }
    joined += call.completion;
  }
  if (joined.length === 0) {
    return undefined;
  }

  const scrubbed = cfg.redaction === "floor" ? redact(joined).text : joined;

  // Raw mode ships the full (redacted) turns for server-side summarization; the
  // standard path caps the excerpt at 320 code points.
  return cfg.sendsFullTurns()
    ? scrubbed
    : truncateChars(scrubbed, EXCERPT_MAX_CHARS);
}

/**
 * Resolve the per-event metadata: `Config` defaults are the base layer, the
 * per-call ambient snapshot is the middle layer, and per-call tags are the top
 * layer (each later layer wins on a shared key). The caps are then applied.
 * Returns `undefined` when the merged map is empty so the wire key is omitted.
 */
export function resolveMetadata(cfg: Config, call: LlmCall): Metadata | undefined {
  const merged: Metadata = {
    ...cfg.metadata,
    ...(call.ambientMetadataSnapshot ?? {}),
    ...call.metadataTags,
  };
  if (Object.keys(merged).length === 0) {
    return undefined;
  }
  const capped = capMetadata(merged);
  return Object.keys(capped).length === 0 ? undefined : capped;
}

/**
 * Assign `key` on `target` only when `value` is defined. Keeps optional wire
 * keys *omitted entirely* when absent (never serialized as `null`).
 */
function setIfPresent<T, K extends keyof T>(
  target: T,
  key: K,
  value: T[K] | undefined,
): void {
  if (value !== undefined) {
    target[key] = value;
  }
}

/** Convert one captured call into a wire event plus its tool-call records. */
function eventFromCall(
  cfg: Config,
  call: LlmCall,
  seq: number,
): [RawEvent, ToolCallWire[]] {
  const sourceRef = `${call.sessionId}::${call.startedAt.getTime()}::${seq}`;
  const srcEventId = sourceEventId(cfg.deviceId, sourceRef);

  const contentExcerpt = buildExcerpt(cfg, call);

  // Required keys first; optionals are added only when present so they don't
  // serialize as null.
  const event: RawEvent = {
    source_event_id: srcEventId,
    ts: rfc3339(call.startedAt),
    kind: call.kind,
    agent: cfg.agent,
    provider: call.provider,
    session_id: call.sessionId,
    tokens: call.tokenUsage,
  };
  setIfPresent(event, "model", call.modelName);
  setIfPresent(event, "cwd", call.cwd);
  setIfPresent(event, "git", call.git);
  setIfPresent(event, "duration_ms", call.durationMs);
  setIfPresent(event, "content_excerpt", contentExcerpt);
  setIfPresent(event, "metadata", resolveMetadata(cfg, call));

  const toolCalls: ToolCallWire[] = call.toolCalls.map((tc, i) => {
    const [argsHash, signatureHash, argsBytes] = hashArgs(tc.args);
    // 16-hex truncation for tool-call ids (vs 32 for events/batches).
    const externalCallId = `tc_${contentHash([srcEventId, String(i)]).slice(0, 16)}`;

    const wireTc: ToolCallWire = {
      external_call_id: externalCallId,
      session_id: call.sessionId,
      source_event_id: srcEventId,
      agent: cfg.agent,
      server: tc.server ?? "builtin",
      name: tc.name,
      call_index: i,
      started_at: rfc3339(tc.startedAt ?? call.startedAt),
      status: tc.status,
      args_hash: argsHash,
      signature_hash: signatureHash,
      args_bytes: argsBytes,
      result_bytes: tc.resultBytes ?? 0,
    };
    setIfPresent(wireTc, "ended_at", tc.endedAt ? rfc3339(tc.endedAt) : undefined);
    setIfPresent(wireTc, "model", call.modelName);
    // command_families: omit entirely when empty, cap at 3.
    const families = (tc.commandFamilies ?? []).slice(0, 3);
    if (families.length > 0) {
      wireTc.command_families = families;
    }
    return wireTc;
  });

  return [event, toolCalls];
}

/**
 * A mutable monotonic sequence counter, passed by reference so per-call dedupe
 * keys stay distinct across flushes within a run (mirrors Rust's `&mut u64`).
 */
export interface SeqRef {
  value: number;
}

/**
 * Drain a batch of captured calls into a wire {@link IngestBatch}. `seqRef` is
 * a monotonic counter bumped once per call to keep per-call dedupe keys
 * distinct within a run.
 *
 * `tool_calls` is omitted entirely from the batch when no call had any (never
 * serialized as `[]`).
 */
export function buildBatch(
  cfg: Config,
  calls: Iterable<LlmCall>,
  seqRef: SeqRef,
): IngestBatch {
  const events: RawEvent[] = [];
  const toolCalls: ToolCallWire[] = [];
  const sourceIds: string[] = [];

  for (const call of calls) {
    seqRef.value += 1;
    const [event, tcs] = eventFromCall(cfg, call, seqRef.value);
    sourceIds.push(event.source_event_id);
    for (const tc of tcs) {
      toolCalls.push(tc);
    }
    events.push(event);
  }

  const batch: IngestBatch = {
    batch_id: batchId(sourceIds),
    device_id: cfg.deviceId,
    daemon_version: cfg.version,
    events,
  };
  // Omit `tool_calls` entirely when empty (additive wire contract).
  if (toolCalls.length > 0) {
    batch.tool_calls = toolCalls;
  }
  // Always send an explicit value reflecting the config so backend usage is
  // off-by-default but users can opt in. `setIfPresent` keeps the wire key
  // omitted only if the flag were ever undefined; the config default makes it a
  // concrete boolean.
  setIfPresent(batch, "auto_taxonomy", cfg.autoTaxonomy);
  return batch;
}
